File: 36-valid-sudoku/valid-sudoku.java
class Solution {
    int getJ(int j){
        if(j<3){
            return 0;
        }
        else if(j<6){
            return 1;
        }else{
            return 2;
        }
    }
    int getBoxIndex(int i,int j){
        if(i<3){
            return getJ(j);
        }
        else if(i<6){
            return 3+getJ(j);
        }else{
            return 6+getJ(j);
        }
    }
    public boolean isValidSudoku(char[][] board) {
        List<HashSet<Character>> Rows=new ArrayList<>();
        List<HashSet<Character>> Cols=new ArrayList<>();
        List<HashSet<Character>> boxes=new ArrayList<>();
        for(int i=0;i<9;i++){
            Rows.add(new HashSet<>());
            Cols.add(new HashSet<>());
            boxes.add(new HashSet<>());
        }
        for(int i=0;i<9;i++){
            for(int j=0;j<9;j++){
                int idx=getBoxIndex(i,j);
                if(board[i][j]=='.'){
                    continue;
                }
                if(Rows.get(i).contains(board[i][j])){
                    System.out.println("Fails here"+i+":"+j);
                    return false;
                }else if(Cols.get(j).contains(board[i][j])){
                    
                    return false;
                }
                else if(boxes.get(idx).contains(board[i][j])){
                    
                    return false;
                }
                else{
                    Rows.get(i).add(board[i][j]);
                    Cols.get(j).add(board[i][j]);
                    boxes.get(idx).add(board[i][j]);
                }
            }
        }
        return true;
    }
}

File: 36-valid-sudoku/valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        
        if(len(board)!=9 or any(len(row)!=9 for row in board)):
            return False

        rows=[[False]*9 for _ in range(9)]
        cols=[[False]*9 for _ in range(9)]
        boxes=[[False]*9 for _ in range(9)]

        for i in range(9):
            for j in range(9):
                if board[i][j]!='.':
                    num=ord(board[i][j])-ord('1')
                    box_index=(i//3)*3+(j//3)
                    if rows[i][num] or cols[j][num] or boxes[box_index][num]:
                        return False
                    rows[i][num]=cols[j][num]=boxes[box_index][num]=True
        return True